### PR TITLE
Use Dataset instead of DataArray for structured xarray data

### PR DESCRIPTION
**Pull request by philsmt**
We've recently begun using structured datatypes for the result of processing data. In particular when the output is of arbitrary length such as for REMI reconstruction and now TIMEPIX centroiding, this is significantly more efficient both in terms of storage space as well as access.

Today a user tipped me off that `xarray.DataArray` basically does not support structured datatypes, making access very complicated and its features almost useless. Instead, the canonical way to handle this in `xarray` appears to wrap it into an `xarray.Dataset` with each field its own variable over common coordinates. For that purpose, I've changed `KeyData.xarray()` to return such a dataset in the case of structured data:

image

Before adding structured data to the test framework, I'd like to discuss whether we can agree on this approach.

### CODE REVIEW

**Comment by takluyver** [comment]
Thanks. This seems reasonable in isolation. A couple of thoughts that came up that might be worth thinking more about:

* I wondered some time ago about a method to load several keys into a dataset object, a bit like the `get_dataframe()` method we have for 1D data. So far it's never seemed all that important to add, but if we ever did want to, it would be more complex if some of keys could already be datasets themselves.
* There's a bigger question over whether we store (on disk & in memory) each 'row' together, as in a structured dtype, or each 'column' together, like xarray would normally work with and like multiple datasets in an HDF5 file. These make different types of access fast, though I think there's a trend towards column-based storage because for a lot of purposes, having similar data together is more efficient. Does this code end up with each variable contiguous in memory, paying the cost of rearranging the data up front? Or does it leave the memory layout as is, potentially making later operations slower? If it leaves the variables interleaved, is there a convenient way to make a column-based variant from it? I just had a quick search and couldn't see an xarray equivalent of `np.ascontiguousarray()`.

**Comment by philsmt** [comment]
> * I wondered some time ago about a method to load several keys into a dataset object, [...] it would be more complex if some of keys could already be datasets themselves.

It is true it will make the implementation more complicated. Conceptually I don't see much of a problem though: from a canonical `xarray` point of view, the individual fields from the structured data type would be variables next to any other external keys. This is actually possible for REMI right now, which has different keys (actually even different structured types, but beside the point) alongside structured data. This is also compatible with the situation we'd have without any structured types at all.

If you'd prefer to not run into this situation and have to handle it, we will need to disable `KeyData.xarray()` for structured data.

> * There's a bigger question over whether we store (on disk & in memory) each 'row' together, as in a structured dtype, or each 'column' together, like xarray would normally work with and like multiple datasets in an HDF5 file. These make different types of access fast, though I think there's a trend towards column-based storage because for a lot of purposes, having similar data together is more efficient. Does this code end up with each variable contiguous in memory, paying the cost of rearranging the data up front? Or does it leave the memory layout as is, potentially making later operations slower? If it leaves the variables interleaved, is there a convenient way to make a column-based variant from it? I just had a quick search and couldn't see an xarray equivalent of `np.ascontiguousarray()`.

As it happens, after posting this thought crossed my mind as well. So far I _think_ we do not guarantee any contiguity of data we return. Naturally it will be the case for some methods, but those shifting detector data may probably not? However, since it's cumbersome to make the individual `DataArray`s in this `Dataset` contiguous, I thought of adding an optional `contiguous=False` keyword argument, which performs the copy operation if desired.
Ultimately your question of which access to optimize for is the entire reason I began using structured data. Karabo hashes and in extension the file format are entirely based around column-based access. This also works nicely with vectorized access for not-to-complicated operations. For the "event-based" data like REMI and Timepix though, the math done per event is complex enough to discourage vectorized access, and the space savings of saving it structured were significant.
For full disclosure, it does make `pasha` more efficient, as well as the C++ level access the user software for REMI has.

**Comment by takluyver** [comment]
Thanks, that sounds good.

Thinking about it again, maybe making the arrays 'column-based' in memory is as simple as calling `dset.copy(deep=True)`, assuming that it copies each variable as a contiguous array. If so, we can avoid the need for an extra parameter.

**Comment by philsmt** [comment]
Then I'll go ahead and add tests?

**Comment by takluyver** [comment]
Sounds good :+1:

**Comment by takluyver** [comment]
Thanks, what's here looks good, but I think you forgot to commit the new file for the `ReconstructedDLD6` mock data class.

File: extra_data/tests/mockdata/proc.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    ])
+
+    def __init__(self, device_id, nsamples=None):
+        super(ReconstructedDLD6, self).__init__(device_id, nsamples=nsamples)
+
+        self.output_channels = ('output/rec',)
+        self.instrument_keys = [
+            ('signals', self.signals_dt, ()),
+            ('hits', self.hits_dt, ()),
+        ]
```
[the remark this answers is not in the document]
Comment (takluyver) [reply extra_data/tests/mockdata/proc.py:34]: Also, I think these should have an extra dimension (in the third item for each key) of 50?

**Comment by philsmt** [comment]
Well spotted, you're right. Thanks!

**Comment by takluyver** [comment]
Did you find any reason not to define the channels and keys on the class? It's not important, just a bit less code that way.

File: extra_data/tests/mockdata/proc.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (philsmt) [reply extra_data/tests/mockdata/proc.py]: Forgot about that, sorry! Fixed.

Small criticism for the current mocking framework: `DeviceBase` assumes `instrument_keys` to be the same for all top-level pipeline keys, which is fairly rarely the case (only ADQ digitizer comes to mind). I don't ultimately need it, but these sources have two top-level keys (`raw` and `rec`) with different keys below. I considered changing it, but ultimately deemed it not necessary for now.

**Comment by philsmt** [comment]
Forgot about that, sorry! Fixed.

Small criticism for the current mocking framework: DeviceBase assumes instrument_keys to be the same for all top-level pipeline keys, which is fairly rarely the case (only ADQ digitizer comes to mind). I don't ultimately need it, but these sources have two top-level keys (raw and rec) with different keys below. I considered changing it, but ultimately deemed it not necessary for now.

**Comment by takluyver** [comment]
Thanks, LGTM.

Yup, sorry about that. I think the digitiser was the example I based it on. The data mocking code was about the first thing I did when I started at XFEL, and it's only had minor changes since then.

**Pull request merged**